ls1intum/Artemis
# [Exam] Programming exercise management

### Checklist

### Motivation and Context
Programming exercise CRUD operations must be adapted for exam mode.

### Description
Client:
- [x] Import for exam programming exercises
- [x] Created new client routes for text-exercise update and creation
For creation: ':examId/exercise-groups/:groupId/programming-exercises/new'
For update: ':examId/exercise-groups/:groupId/programming-exercises/:exerciseId/edit'
For import: ':examId/exercise-groups/:groupId/programming-exercises/import/:exerciseId'
- [x] Load programming exercise module
- [x] Deactivate categories in exam mode for now, as this relies on a Course
- [x] Added "Add programming exercise" button to exercise group component  
- [x] Added exam mode to ProgrammingExerciseDetail Component. Edit should work on this page now.

Server:
- [x] Import functionality supports exam exercises now
- [x] Deactivate Tutor/Student notification on exercise creation/update
- [x] Add additional authorization check to GET programming exercise route. Only instructors and admins should be allowed to access an exam programming exercise
- [x] Adapted authorization check for deletion
- [x] Deactivated hints, test-cases, results-with-feedbacks and save problem statement routes for exam exercises. These routes are called by the ProgrammingExerciseEditableInstructionComponent, if the user tries to edit the exercise.

**There is a lot more functionality tied to programming exercises. For example actions available on the view page. These functions are not in the scope of this PR and will be adapted in a follow up PR.**

### Steps for Testing
_Ensure that the existing CRUD operations for regular programming exercises still work_

1. Log in as admin or instructor
2. Create a regular programming exercise
--> Exercise was created
3. Update the exercise
--> Exercise was updated
4. View the exercise
--> Can open the page
5. Delete the exercise
--> Exercise was deleted

_Ensure that CRUD operations for exam programming exercises work_

1. Log in as admin or instructor
2. Create an exam
3. Create an exercise group
4. Create a programming exercise
--> Exercise was created
5. Edit the programming exercise
--> Update was saved
6. Navigate to `/course-management/:courseId/programming-exercises/:exerciseId` using the id of an exam programming exercise
--> Should be able to see the exercise
7. Delete the exercise
--> Exercise was deleted

_Test import functionality of programming exercises_
1. Create a normal exercise -> A
2. Create a programming exercise in an exam -> B
3. Navigate to the ExerciseGroup view and click on the import new programming exercise button
--> Modal should show up with a list of programming exercises for which the user has access rights (instructor, admin)
4. Choose A for import
--> Import page should open
5. Insert the short name of A, fill out the other mandatory fields and click on import
--> Message should be displayed the the short name already exists
6. Choose an unique short name and click on the import button
--> Exercise should be imported
7. Navigate to the course exercises management page
8. Click on import new programming exercise
--> Ensure the import works by going through the steps above 

_Ensure correct authorization for programming exercises_
1. Log in as a tutor or student
2. Navigate to `course-management/:courseId/exams/:examId/exercise-groups/:groupId/programming-exercises/new`
--> You should not be able to access the route
4. Use the id of an exam exercise for the following steps
5. Navigate to `course-management/:courseId/exams/:examId/exercise-groups/:groupId/programming-exercises/:exerciseId/edit`
--> You should not be able to access the route
7. Navigate to `/course-management/:courseId/programming-exercises/:exerciseId`
--> You should not be able to see the exercise

### Screenshots
